src/libcore/iter/mod.rs

```diff
@@ -1643,20 +1708,145 @@ impl<I: DoubleEndedIterator, U, F> DoubleEndedIterator for FlatMap<I, U, F> wher
 ///
 /// [`fuse()`]: trait.Iterator.html#method.fuse
 /// [`Iterator`]: trait.Iterator.html
-#[derive(Clone, Debug)]
 #[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct Fuse<I> {
-    iter: I,
-    done: bool
+    iter: <I as SpecFuse>::T,
```

Comment:
I'll be curious to see your approach, as I earlier reported another instance of this issue here: https://github.com/rust-lang/rust/issues/30642#issuecomment-180879882